### DISCUSSION TITLE
An error will be reported when the archive contains Chinese

### DISCUSSION THREAD

**Discussion by ja-mitiankai** [rikyoz/bit7z#136] [Q&A] [answered]
The error is 'Failed to get stream'. I don't know how to solve this problem. Can someone help me. Here is my code: 

`
using namespace bit7z;
Bit7zLibrary lib("7z.dll");
BitFileExtractor extractor(lib, BitFormat::Zip);
extractor.extract(archive_path, out_dir);
`

**Comment by rikyoz** [comment]
Hi!
Unfortunately, I cannot replicate the issue.
Could you give some more details about it?
Are the Chinese characters in the archive name or the name of a file/folder contained in the archive?
Which version of Windows and 7-zip are you using?
Are you sure the problem is due to the Chinese characters?
All possible details would be helpful!
Thank you! 🙏

**Reply by ja-mitiankai** [reply]
Hi!
First, thank you very much for your prompt recovery.

> Are the Chinese characters in the archive name or the name of a file/folder contained in the archive?
The name of the file/folder in the archive contains Chinese, not the archive itself.
For example, such archive structures can cause errors: 
aa.zip
    -- 水文化生态园桥.txt
or:
aa.zip
     -- 水文化生态园桥
                    -- -- bb.txt
or: 
水文化生态园桥.zip
     -- 水文化生态园桥.txt
This will execute correctly: 
水文化生态园桥.zip
        -- aa.txt
or:
水文化生态园桥.zip
        -- aa
            -- -- bb.txt
or:
aa.zip
    -- xx.txt

>Which version of Windows and 7-zip are you using?
Windows10 and 7-zip 22.01

>Are you sure the problem is due to the Chinese characters?
Yes, I'm sure. But I'm not sure if only Chinese would cause such an error.
An error will be reported when the file in the archivecontains such a name. Some Chinese characters can be successfully decompressed, but there may be garbled code.
This name will report an error: 
For example: "uuu.zip/水文化生态园桥.txt"
But after I change the above name, it will execute it correctly.
Changed name: "uuu.zip/aa.txt"

**Reply by rikyoz** [reply]
> Hi! First, thank you very much for your prompt recovery.

You're welcome!

> > Are the Chinese characters in the archive name or the name of a file/folder contained in the archive?
> > The name of the file/folder in the archive contains Chinese, not the archive itself.
> > For example, such archive structures can cause errors:
> > aa.zip
> > -- 水文化生态园桥.txt
> > or:
> > aa.zip
> > -- 水文化生态园桥
> > -- -- bb.txt
> > or:
> > 水文化生态园桥.zip
> > -- 水文化生态园桥.txt
> > This will execute correctly:
> > 水文化生态园桥.zip
> > -- aa.txt
> > or:
> > 水文化生态园桥.zip
> > -- aa
> > -- -- bb.txt
> > or:
> > aa.zip
> > -- xx.txt
> 
> > Which version of Windows and 7-zip are you using?
> > Windows10 and 7-zip 22.01
> 
> > Are you sure the problem is due to the Chinese characters?
> > Yes, I'm sure. But I'm not sure if only Chinese would cause such an error.
> > An error will be reported when the file in the archive contains such a name. Some Chinese characters can be successfully decompressed, but there may be garbled code.
> > This name will report an error:
> > For example: "uuu.zip/水文化生态园桥.txt"
> > But after I change the above name, it will execute it correctly.
> > Changed name: "uuu.zip/aa.txt"

Thanks for the further details!
Unfortunately, I'm not able to replicate your issue.
I tested some archives structured as you described, and bit7z is fine with extracting them.

What compiler are you using?
If you are using MSVC, did you enable UTF-8 encoding of the source and executable character sets (https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170)?

By default, bit7z uses `std::string` for paths and considers them UTF-8 encoded; hence your issue if they're not.
Alternatively, you could enable the `BIT7Z_USE_NATIVE_STRING` flag while compiling bit7z: this will set the library to use UTF-16 encoded strings (`std::wstring`).
In this case, you must use `L"..."` string literals for the paths.

**Reply by ja-mitiankai** [reply]
Thank you very much for your reply and the method you provided.

> What compiler are you using?
I am using Visual Studio 2019 (v142). MSVC is 14.29.30133.

I tried to set UTF-8, but the decompression failed. I also tried using BIT7Z_ USE_ NATIVE_ STRING flag compiling bit7z, but its link failed(LNK2001). I can execute the initial compiled bit7z, but decompression failed. I don't know if there was a problem with my bit7z compilation process or something else right now. I'm not very good at C++, it's too complicated for me. Can you help me compile bit7z and share it with me?

**Reply by rikyoz** [reply, marked as answer]
I'm sorry for not getting back to you sooner!

> > What compiler are you using?
> > I am using Visual Studio 2019 (v142). MSVC is 14.29.30133.

Uhm, ok, thanks!

> I tried to set UTF-8, but the decompression failed. 

Uhm, strange! I have some guesses for this, but I think it's better to check the issues with the `BIT7Z_USE_NATIVE_STRING` flag before, at least for now.

> I also tried using BIT7Z_ USE_ NATIVE_ STRING flag compiling bit7z, but its link failed(LNK2001). I can execute the initial compiled bit7z, but decompression failed. I don't know if there was a problem with my bit7z compilation process or something else right now. I'm not very good at C++, it's too complicated for me. Can you help me compile bit7z and share it with me?

The link failure is probably due to the fact that you need to define the flag in your program, too; otherwise, you end up having a mismatch between the API declared in the bit7z headers and the one implemented in the compiled library.

If this is the case, adding 
```cpp
#define BIT7Z_USE_NATIVE_STRING
```
just before your `#include` of bit7z should work, e.g.,
```cpp
#define BIT7Z_USE_NATIVE_STRING
#include <bit7z/bit7zlibrary.hpp>
#include <bit7z/bitfileextractor.hpp>
```

In any case, you can build the library as follows
```bash
cd <bit7z folder>
mkdir build && cd build
cmake ../ -DCMAKE_BUILD_TYPE=Release -DBIT7Z_USE_NATIVE_STRING=ON
cmake --build . -j --config Release
```
This will create the `.lib` file in `<bit7z folder>/lib/x64` (or `<bit7z folder>/lib/x86` for 32-bit builds).

Alternatively, if your project uses CMake, you can simply add this code to your `CMakeLists.txt`:
```cmake
add_subdirectory( ${CMAKE_SOURCE_DIR}/third_party/bit7z ) # this is the path where you put the bit7z source code
set(BIT7Z_USE_NATIVE_STRING ON CACHE BOOL "Enable native strings" FORCE)
target_link_libraries( ${TARGET_NAME} PRIVATE bit7z64 ) # or bit7z on x86
```

Please, let me know if any of these options work!

**Reply by ja-mitiankai** [reply]
Thank you very much for your help. I have succeeded.
I set it up last time -DBIT7Z_USE_NATIVE_STRING=true, it didn't work. Originally, it was set to "ON". I saw that the CMakeCache.txt file is all "OFF", and I didn't try to set it to "ON". It's my problem, I'm sorry for wasting some of your time. I was foolish.

**Reply by rikyoz** [reply]
> Thank you very much for your help. I have succeeded. I set it up last time -DBIT7Z_USE_NATIVE_STRING=true, it didn't work. Originally, it was set to "ON". I saw that the CMakeCache.txt file is all "OFF", and I didn't try to set it to "ON". It's my problem, I'm sorry for wasting some of your time. I was foolish.

No worries, these things happen! I'm happy to have helped!
